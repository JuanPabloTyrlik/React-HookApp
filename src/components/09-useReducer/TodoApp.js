import React, { useReducer, useEffect } from 'react';
import './styles.css';
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm';

const TodoApp = () => {
    const init = () => {
        return JSON.parse(localStorage.getItem('TodoList')) || [];
    };

    const [todoList, dispatch] = useReducer(todoReducer, [], init);

    const [{ description }, handleInputChange, reset] = useForm({
        description: '',
    });

    useEffect(() => {
        localStorage.setItem('TodoList', JSON.stringify(todoList));
    }, [todoList]);

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 0) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        };

        const action = {
            type: 'Add',
            payload: newTodo,
        };

        dispatch(action);

        reset();
    };

    const handleDelete = (todo) => {
        const action = {
            type: 'Delete',
            payload: todo,
        };
        dispatch(action);
    };

    const handleToggle = (todo) => {
        dispatch({
            type: 'Toggle',
            payload: todo,
        });
    };

    return (
        <>
            <h1>TodoApp ( {todoList.length} )</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <ul className="list-group list-group-flush">
                        {todoList.map((todo, i) => (
                            <li key={todo.id} className="list-group-item">
                                <p 
                                    className={`${todo.done && 'complete'}`}
                                    onClick={() => handleToggle(todo)}
                                >
                                    {i + 1}. {todo.desc}
                                </p>
                                <button
                                    onClick={() => handleDelete(todo)}
                                    className="btn btn-danger"
                                >
                                    Borrar
                                </button>
                            </li>
                        ))}
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Agregar TODO</h4>
                    <hr />
                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            name="description"
                            className="form-control"
                            placeholder="Aprender..."
                            autoComplete="off"
                            value={description}
                            onChange={handleInputChange}
                        />
                        <button
                            type="submit"
                            className="btn btn-outline-primary mt-1 btn-block"
                        >
                            Agregar
                        </button>
                    </form>
                </div>
            </div>
        </>
    );
};

export default TodoApp;
