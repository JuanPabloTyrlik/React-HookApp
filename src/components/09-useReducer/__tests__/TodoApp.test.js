import { act } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import React from 'react';
import TodoApp from '../TodoApp';

describe('Tests on TodoApp', () => {
    const demoTodos = [
        {
            id: 1,
            desc: 'Aprender React',
            done: true,
        },
        {
            id: 2,
            desc: 'Aprender Mongo',
            done: false,
        },
    ];
    Storage.prototype.setItem = jest.fn();
    test('should render correctly', () => {
        const wrapper = shallow(<TodoApp />);
        expect(wrapper).toMatchSnapshot();
    });
    test('should function correctly', () => {
        const wrapper = mount(<TodoApp />);
        act(() => {
            wrapper.find('TodoAddTask').prop('handleAdd')(demoTodos[0]);
            wrapper.find('TodoAddTask').prop('handleAdd')(demoTodos[1]);
        });
        expect(wrapper.find('h1').text()).toMatch('TodoApp ( 2 )');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);

        act(() => {
            wrapper.find('TodoList').prop('handleDelete')(demoTodos[1]);
        });
        expect(wrapper.find('h1').text()).toMatch('TodoApp ( 1 )');
        expect(localStorage.setItem).toHaveBeenCalledTimes(3);
    });
});
