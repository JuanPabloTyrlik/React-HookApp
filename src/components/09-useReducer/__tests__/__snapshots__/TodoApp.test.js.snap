// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on TodoApp should render correctly 1`] = `
<Fragment>
  <h1>
    TodoApp ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todoList={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAddTask
        handleAdd={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
